### DISCUSSION TITLE
Allow unassigned rule calls after an assignment

### DISCUSSION THREAD

**Discussion by spoenemann** [eclipse-langium/langium#205] [Ideas]
Xtext forbids an unassigned rule call after an assignment because it uses the assignment as trigger for creating an EClass instance. We could relax this, and with that would gain a big advantage when it comes to factoring out common prefixes.

Example:
```
Entry:
    A | B | C;

A:
    prefix=Prefix
    'a' ...;

B:
    prefix=Prefix
    'b' ...;

C:
    prefix=Prefix
    'c' ...;
```
The `Entry` rule requires a potentially long lookahead (depending on the `Prefix`) rule, which can be bad for the parser performance. An alternative is to write it like this:
```
Entry:
    prefix=Prefix content=Content;

Content:
    A | B | C;

A:
    'a' ...;

B:
    'b' ...;

C:
    'c' ...;
```
The downside is that the AST then has two separate objects for `Entry` and `Content`, while the original version had a single object where `Entry` is a supertype of `A`, `B` and `C`. New idea:
```
Entry:
    prefix=Prefix
    (A | B | C);

A:
    'a' ...;

B:
    'b' ...;

C:
    'c' ...;
```
This would first parse the `Prefix` and then reassign the object `$type` to either `A`, `B` or `C` without creating a separate object. We have extracted the common prefix without modifying the AST.

This has limitations:
```
Foo:
    A B C;
```
What shall we do here? Parse A, then reassign the type to B, and then again to C? We could constrain unassigned rule calls to a single set of alternatives per rule.

**Comment by pluralia** [comment]
Currently, there is a workaround for Xtext and Langium:
```
Entry:
    Prefix (
        {infer A.prefix=current} 'a' ...
    |   {infer B.prefix=current} 'b' ...
    |   {infer C.prefix=current} 'c' ...
    );
```

Such grammar generates:
```
export type Entry = A | B | C | Prefix;

export interface A extends AstNode {
    prefix: Prefix
    ...
}

// similar for B and C
```

This syntax imitate behavior of the proposed syntax, however, it allows to grab the only 1 common element `prefix: Prefix`. There is no replacement for the case of 2 common elements like here, where `A`, `B` and `C` have properties `prefix: Prefix` and `anotherPrefix: AnotherPrefix`.
```
Entry:
    prefix=Prefix anotherPrefix=AnotherPrefix (A | B | C);
```
